javalin/src/main/java/io/javalin/plugin/PluginManager.kt

```diff
@@ -1,19 +1,56 @@
 package io.javalin.plugin
 
 import io.javalin.config.JavalinConfig
+import io.javalin.util.parentClass
+import kotlin.reflect.KClass
+import kotlin.reflect.full.isSubtypeOf
+import kotlin.reflect.full.isSupertypeOf
+import kotlin.reflect.jvm.jvmErasure
+import kotlin.reflect.typeOf
 
 class PluginManager internal constructor(private val cfg: JavalinConfig) {
 
     private val plugins = mutableListOf<Plugin<*>>()
     private val initializedPlugins = mutableListOf<Plugin<*>>()
     private val enabledPlugins = mutableListOf<Plugin<*>>()
+    private val contextPlugins = mutableMapOf<PluginKey<*>, ContextExtendingPlugin<*, *>>()
```

Comment:
Intuitively it doesn't make sense to me that these live in different collections, why is it required?

Reply:
All plugins get added to `enabledPlugins`; only plugins that we need to lookup by key get added to `contextPlugins`.  We *could* store everything in a map, but then we need to generate throwaway plugin keys for all the "regular" plugins, which doesn't seem like a win.